### PR TITLE
Feature/disks

### PR DESCRIPTION
**Pull request by Mpdreamz**
## Disk size 
This adds the ability to choose disksize (Small, Medium, Large) which corresponds with P10, P20, P30 in the premium storage tier. If you do not need 1TB disks this allows you to attach smaller disks. It is important to note that smaller disks come with less performance too see:

image

## Number of disks 

It also adds the abillity to specify the number of disks to configure in raid. Each VM has a different maximum number of disks and you can pass any smaller number. Take for example this listing for DS_v2 found [on this overview page](https://docs.microsoft.com/en-us/azure/virtual-machines/virtual-machines-linux-sizes?toc=%2fazure%2fvirtual-machines%2flinux%2ftoc.json#dsv2-series)

image

The default number of disks already corresponds to this listing and now its also the per VM size max you can specify. If you do specify a value greater then the vm max it will min() to the vm max.

## No mdadm raid for single disks 

If you specify one disk, we omit the the raiding completely now and just mount the attached disk

## Specifying `0` disks

`0` is also a valid option and will cause the ephemeral disk to be used. This is not intended for long running clusters but can really help keeping the costs down for short lived ones. If you run fully replicated with tons of backups and are aware of the risks this could also be an option for the budget constraint projects.

`nodesPerStorageAccount` takes these two variables into account now as well and no longer defaults to `10` in all situations. 

`maxOf(storage_type, diskSize) / numberOfDataDisks`

`maxOf` will yield `40` for all standard storage accounts and for premium the following:

* Small: 250
* Medium: 60
* Large: 34

This means we will create less storage accounts and saturate them better,  an azure subscription by default only has 200 storage accounts they can create.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -260,7 +260,7 @@ azure group create <name> <location>
 Next we can either use our published template directly using `--template-uri`
 
 ```sh
-azure group deployment create --template-uri https://raw.githubusercontent.com/elastic/azure-marketplace/master/src/mainTemplate.json --parameters-file parameters/password.parameters.json -g name
+azure group deployment create --template-uri https://raw.githubusercontent.com/elastic/azure-marketplace/feature/disks/src/mainTemplate.json --parameters-file parameters/password.parameters.json -g name
```
Comment (russcam) [inline README.md]: Need to run `npm run links -- --branch master` before merging into `master`

**Comment by Mpdreamz** [comment]
As per the suggestion of @niemyjski (:+1:) this now also allows you to downgrade from premium to standard storage by overriding `storageAccountType` in maintemplate.

**Comment by russcam** [comment]
I've gone through this and added a few further commits to:

1. Add data disk parameters to the README
2. Constrain data disk count values to `0`, `1`, `2`, `4`, `8`, `16`, `32`, `40` with a default value of `40`.

   Specifying any other value other than above fails the deployment.

    I've converted this parameter to a `string` in order to work correctly with `allowedValues` and a `defaultValue`; when the type is `int`, template web deploy renders the parameter with an empty default value and even removing the `defaultValue` (i.e. making it required), still renders the parameter with an empty default value. Looks like a bug in the portal (I'll file an issue).

3. Add `minValue` validation for data and client nodes

@Mpdreamz, one thing left to discuss is the `Default`, `Standard` and `Premium` options for `storageAccountType`. Since `Premium` is not a valid value for SKUs that don't support Premium disks, I think we should remove it as an option to prevent invalid parameter combinations, and ensure that we document that `Default` will be Premium disks for SKUs that support Premium. Thoughts?

**Comment by russcam** [comment]
> I've converted this parameter to a string in order to work correctly with allowedValues and a defaultValue; when the type is int, template web deploy renders the parameter with an empty default value and even removing the defaultValue (i.e. making it required), still renders the parameter with an empty default value. Looks like a bug in the portal (I'll file an issue

On further thought, I think this is better to be kept as `int` and let the portal get fixed; `int` is the type that it should be from the CLI/PowerShell

**Comment by Mpdreamz** [comment]
> Since Premium is not a valid value for SKUs that don't support Premium disks, I think we should remove it as an option to prevent invalid parameter combinations, and ensure that we document that Default will be Premium disks for SKUs that support Premium. Thoughts?

👍 this is the more elegant fail safe solution

File: src/mainTemplate.json [hunk 2]
```diff
@@ -249,17 +249,17 @@
       }
     },
     "vmDataDiskCount": {
-      "type": "string",
-      "defaultValue": "40",
+      "type": "int",
```
Comment (Mpdreamz) [inline src/mainTemplate.json]: We can also revert this https://github.com/elastic/azure-marketplace/pull/84/commits/c7fa33b401e1b2cbb25231cff00abe1d0a869e71 now that this is an int again.

**Pull request merged**